src/transforms/kv_parser.rs

```diff
@@ -0,0 +1,404 @@
+    event::{self, Event},
+    topology::config::{DataType, TransformConfig, TransformContext, TransformDescription},
+    types::{parse_conversion_map, Conversion},
+};
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::str;
+use string_cache::DefaultAtom as Atom;
+
+#[derive(Debug, Default, Derivative, Deserialize, Serialize)]
+#[serde(default, deny_unknown_fields)]
+pub struct KeyValueConfig {
+    #[derivative(Default(value = "true"))]
+    pub drop_field: bool,
+    pub field: Option<Atom>,
+    #[derivative(Default(value = " "))]
+    pub field_split: String,
+    #[derivative(Default(value = "true"))]
+    pub overwrite_target: bool,
+    pub separator: String,
+    pub target_field: Option<Atom>,
+    pub trim_key: Option<String>,
+    pub trim_value: Option<String>,
+    pub types: HashMap<Atom, String>,
+}
+
+inventory::submit! {
+    TransformDescription::new::<KeyValueConfig>("kv_parser")
+}
+
+#[typetag::serde(name = "kv_parser")]
+impl TransformConfig for KeyValueConfig {
+    fn build(&self, _cx: TransformContext) -> crate::Result<Box<dyn Transform>> {
+        let field = self
+            .field
+            .as_ref()
+            .unwrap_or(&event::log_schema().message_key());
+
+        let conversions = parse_conversion_map(&self.types)?;
+
+        Ok(Box::new(KeyValue {
+            conversions,
+            drop_field: self.drop_field,
+            field: field.clone(),
+            field_split: self.field_split.clone(),
+            overwrite_target: self.overwrite_target,
+            separator: self.separator.clone(),
+            target_field: self.target_field.clone(),
+            trim_key: self.trim_key.clone(),
+            trim_value: self.trim_value.clone(),
+        }))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn transform_type(&self) -> &'static str {
+        "kv_parser"
+    }
+}
+
+pub struct KeyValue {
+    conversions: HashMap<Atom, Conversion>,
+    drop_field: bool,
+    field: Atom,
+    field_split: String,
+    overwrite_target: bool,
+    separator: String,
+    target_field: Option<Atom>,
+    trim_key: Option<String>,
+    trim_value: Option<String>,
+}
+
+fn kv_parser(
+    pair: String,
+    field_split: &String,
+    trim_key: &Option<String>,
+    trim_value: &Option<String>,
+) -> Option<(Atom, String)> {
+    let pair = pair.trim();
+    let mut field;
+
+    if field_split.is_empty() {
+        let mut kv_pair = pair.split_whitespace();
+        let key = kv_pair.nth(0)?;
+        let val = kv_pair.nth(1)?;
+        let count = kv_pair.count();
+
+        if count < 2 {
+            return None;
+        } else if count > 2 {
+            error!(
+                message = "KV parser expected 2 values, but got {count}",
+                count = count,
+                rate_limit_secs = 30
+            );
+            return None;
+        }
```

Comment:
```suggestion
        let key = kv_pair.next()?;
        let val = kv_pair.next()?;
        if let kv_pair.next().is_some() {
            error!(
                message = "KV parser saw more than one separator",
                rate_limit_secs = 30
            );
            return None;
        }
```

Reply:
This is what I came here for, good suggestions by more experienced rust devs!

Couple questions:
- is `next()` more performant or idiomatic than `nth()`, or just better in this case because it eliminates the need for `count`?
- the error case of more than one separator _should_ be impossible here but, I figured it would be safest to write a catch for it. The other fail case of having no separator would be possible and more common. Should we add an `error` for that so it's logged?

Reply:
I wouldn't expect a major performance difference in this case. However, `next()` is the underlying iterator method, and `nth()` is a wrapper, so I don't think the latter can produce _better_ code in this case (might not be true for iterators over actual arrays or ranges). In terms of eliminating the count, that eliminates the need to continue the iteration beyond the next match.

I don't see the case of more than one separator being eliminated through code constructs, so it has to be handled. Even if you think it should never appear in the input, it is still a possibility.

It is probably worth answering the question if `key=value=another` should be an error or parsed as `key` and `value=another`.

Reply:
I think I would prefer if this to be parsed `key: "value=another"`. There are arguments I can think of on both sides. I don't have really strong feelings either way.